File: 13/d13-2.f90
program d13_2
  use d13_helper, only : build_tree, compare_trees, node_types, tree_node
  implicit none

  integer, parameter :: max_line_len = 1024

  integer :: fd, iostat, irecord, idivpack
  character(max_line_len) :: line
  type(tree_node), allocatable :: tree
  type(tree_node) :: divpackages(2)
  integer :: divpackinds(2)

  divpackages(:) = [&
      & tree_node(nodetype=node_types%listnode,&
      & child=tree_node(nodetype=node_types%listnode,&
      & child=tree_node(nodetype=node_types%intnode, intval=2))),&
      &&
      & tree_node(nodetype=node_types%listnode,&
      & child=tree_node(nodetype=node_types%listnode,&
      & child=tree_node(nodetype=node_types%intnode, intval=6)))&
      & ]
  divpackinds(:) = [1, 2]

  open(newunit=fd, file="input.dat")
  linereader: do
    do irecord = 1, 2
      read(fd, "(a)", iostat=iostat) line
      if (iostat /= 0) exit linereader
      call build_tree(trim(line), tree)
      do idivpack = 1, size(divpackages)
        if (compare_trees(tree, divpackages(idivpack)) == -1) then
          divpackinds(idivpack) = divpackinds(idivpack) + 1
        end if
      end do
    end do
    read(fd, *, iostat=iostat)
    if (iostat /= 0) exit
  end do linereader
  close(fd)

  print "(a, i0)", "Encoder key for distress signal: ", product(divpackinds)

end program d13_2

File: 13/d13-2.py
import json

def compare(lhs, rhs):
    """Compares two lists"""
    if isinstance(lhs, int) and isinstance(rhs, int):
        diff = lhs - rhs
        return diff // abs(diff) if diff else 0
    if isinstance(lhs, list) and isinstance(rhs, int):
        res = compare(lhs, [rhs])
        return res
    if isinstance(lhs, int) and isinstance(rhs, list):
        res = compare([lhs], rhs)
        return res
    for ll, rr in zip(lhs, rhs):
        comp = compare(ll, rr)
        if comp:
            return comp
    diff = len(lhs) - len(rhs)
    return diff // abs(diff) if diff else 0

packages = []
fp = open("input.dat", "r")
while True:
    line1 = fp.readline().strip()
    line2 = fp.readline().strip()
    fp.readline()
    if not line1 or not line2:
        break
    packages += [json.loads(line1), json.loads(line2)]
fp.close()

divpackages = [[[2]], [[6]]]
divpackinds = [1, 2]
for package in packages:
    for idivpack, divpack in enumerate(divpackages):
        if compare(package, divpack) == -1:
            divpackinds[idivpack] += 1

print(f"Encoder key for distress signal: {divpackinds[0] * divpackinds[1]}")
